### ISSUE TITLE
Using different (external) dataset with different image resolution

### ISSUE THREAD

**Issue by rahul-nt** [mitroadmaps/roadtracer#11]
Hi
I'm trying to run the code for the road extraction dataset from DeepGlobe challenge, where the images are 1024x1024x3, and (even though from 3 different regions but) have been named independent of the regions. It seems to me that the code however is written in accordance with the data acquired from the google maps. Are there any instructions regarding as to with what changes the code should be run to make it work for any external dataset?

**Comment by rahul-nt** [comment]
@uakfdotb 
Thank you so much for the reply. I will try this out.

**Comment by rahul-nt** [comment]
Hi
I tried to look at the things more closely after reply. If I understand correctly, the roadtracer code creates the labels/masks (the ground truth) for the training images by running the go scripts in the dataset directory.
But what if we have already have the satellite images as well the mask images [the ground truth] (as it is in the case of DeepGlobe's dataset)? Because, from your reply, what I understood is that the graph files as well as the json files are important for running the training and eventually the prediction.

**Comment by uakfdotb** [comment]
The roadcnn code (not RoadTracer system, but rather a baseline from the paper) uses the masks produced by dataset/5_truth_tiles.go, but RoadTracer doesn't actually use the masks at all, only the graphs.

So if you only have masks, you could use them directly with roadcnn, but not with roadtracer.

utils/mapextract.py will take an image and produce a graph from it via thresholding (binary mask from 8-bit image given a threshold), morphological thinning (thin foreground regions until everything is single-pixel-wide lines), and extracting a graph by following the remaining foreground pixels.

This utility is included to convert outputs from roadcnn to vector graph file format. But you could use it to convert the DeepGlobe masks to ground truth graphs as well.

One thing to keep in mind is that RoadTracer never segments the imagery, it produces a graph directly. So you won't be able to compare it on DeepGlobe's intersection-over-union metric. This is a bad metric anyway because it only looks at the segmentation output, not the actual connectivity / topology of the graph (e.g., even if you have many disconnections your performance on metric might still be good, but the map won't be usable for any practical application).

**Comment by uakfdotb** [comment]
Those two modifications sound good. You would directly create the ground truth road network graphs in pixel coordinates based on DeepGlobe dataset, right?

You will have to make some similar adjustments to roadtracer/tileloader.py. Inference requires one point known on the road network, for now maybe split the training set from DeepGlobe into train and test set. But later you could apply on DeepGlobe test set by using peaks in the segmentation output (from another model, e.g. roadcnn) as starting locations for RoadTracer.

**Comment by rahul-nt** [comment]
Is it possible to run roadtracer inference on DeepGlobe data, using the pre-trained model and graphs from this link:
https://roadmaps.csail.mit.edu/roadtracer/

I tried looking infer.py and tileloader.py, and I'm not sure where to start from to makes changes to accomplish the above.
Also, is the documentation for the code available?

**Comment by anil-2185** [comment]
> anil-2185 Those two modifications sound good. You would directly create the ground truth road network graphs in pixel coordinates based on DeepGlobe dataset, right?
> 
> You will have to make some similar adjustments to roadtracer/tileloader.py. Inference requires one point known on the road network, for now maybe split the training set from DeepGlobe into train and test set. But later you could apply on DeepGlobe test set by using peaks in the segmentation output (from another model, e.g. roadcnn) as starting locations for RoadTracer.

Thanks @uakfdotb for response. 
I am working on the modifications to train the roadtracer on DeepGlobe. As suggested I am able to run the code on Deepglobe, but has few doubts:

-  I am not able to understand the subtraction of number 256 in "4_startlocs.go". But found that subtraction is considered necessary in tileloader.py and train.py. Is it required to set the path (start and end rectangle) in the center of image?

- Window size (search area, I assume) is 512 in tileloader.py but in train.py is 256 and model also expects input of size (?,256,256,5). should the window size in tileloader.py be also 256?

- As in DeepGlobe, testing tiles can be large, as compare to single testing tile of Chicago. Can you please suggest the changes for the same in train.py?

Here are few input images to model for training for the verification.
![1_9_3_overlay](https://user-images.githubusercontent.com/7984207/48731444-0a170b80-ec63-11e8-8a7b-d37337544474.png)
![1_9_2_overlay](https://user-images.githubusercontent.com/7984207/48731445-0a170b80-ec63-11e8-8658-6a807d8f64c5.png)
![1_9_1_overlay](https://user-images.githubusercontent.com/7984207/48731447-0a170b80-ec63-11e8-8506-44bebe3ebdff.png)
![1_9_0_overlay](https://user-images.githubusercontent.com/7984207/48731448-0aafa200-ec63-11e8-9307-90da73f51fa6.png)

**Comment by uakfdotb** [comment]
> I am not able to understand the subtraction of number 256 in "4_startlocs.go". But found that subtraction is considered necessary in tileloader.py and train.py. Is it required to set the path (start and end rectangle) in the center of image?

That code is to make sure there are enough starting locations in the tile to use it for training. Starting locations near the border of the image will be excluded because during training we take crops of the imagery and there is no code in place to allow crops that go off of the image canvas. You could modify the training/inference code to allow these crops and set the pixels to 0 if they are outside the image bounds.

> Window size (search area, I assume) is 512 in tileloader.py but in train.py is 256 and model also expects input of size (?,256,256,5). should the window size in tileloader.py be also 256?

It should be okay to set it to 256, it just needs to be larger than the actual window size.

> As in DeepGlobe, number of testing tiles can be more than 1, as compare to single testing tile of Chicago. Can you please suggest the changes for the same in train.py? I tried changing the code similar to train_tiles and create list of test_tiles but I am struggling to run the testing stage...any help is appreciated.

For DeepGlobe you should make each image a separate region, and it can be BLAH_0_0.png.

> How to visualize the graph over the image, as I am not able to get any clue to change the CreateSVG method in Common package?

It may be easier to write your own code. You can see discoverlib/graph.py to load the graph, and then the vertex.point Point objects just reference pixel positions in the imagery. So you can draw lines along edges using your desired programming language / library.

> While inferring only single starting location is randomly chosen from the ground truth (consider for validation set). Thus I am unable to infer the road network completely. How did you rectify this while testing? an example image is shown below with ground truth and inferred roads starting with 2 random locations.

The most straightforward way would be to first run a segmentation model, and then take maximum value in the segmentation output as starting location.

**Comment by uakfdotb** [comment]
More on using segmentation approach:

After running segmentation model, use the maximum value pixel in the segmentation output as starting location for RoadTracer. Then, run RoadTracer to termination. Next, zero out all pixels in the segmentation output that are close to any edge in the output of RoadTracer so far. If the maximum value pixel is above a certain threshold (e.g. 0.5), then repeat the process.

This way, if there are two sets of disconnected roads in the image, we can run RoadTracer on both sides of the road network.
